flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointMetadataOutputStream.java

```diff
@@ -162,4 +164,97 @@ public FsCompletedCheckpointStorageLocation closeAndFinalizeCheckpoint() throws
             }
         }
     }
+
+    static MetadataOutputStreamBackend getOutputStreamBackend(
+            final FileSystem fileSystem, final Path metadataFilePath) throws IOException {
+        try {
+            RecoverableWriter recoverableWriter = fileSystem.createRecoverableWriter();
+            if (fileSystem.exists(metadataFilePath)) {
+                throw new IOException("Target file " + metadataFilePath + " is already exists.");
+            }
+            return new RecoverableStreamBackend(recoverableWriter.open(metadataFilePath));
+        } catch (Throwable throwable) {
+            LOG.warn("Errors on creating recoverable writer.", throwable);
+        }
+        return new FSDataOutputStreamBackend(fileSystem, metadataFilePath);
+    }
+
+    /** The backend to wrap a whole output stream. */
+    interface MetadataOutputStreamBackend {
+
+        FSDataOutputStream getOutput();
+
+        void commit() throws IOException;
+
+        void close() throws IOException;
+
+        void finalize() throws IOException;
```

Comment:
`finalize()` sounds like we will call it in the end (compared with `CheckpointMetadataOutputStream#closeAndFinalizeCheckpoint`). I think method such as `abort()` could be better.

Reply:
Make sense.